Add flag to control reboot interval

hakonhall opened the pull request:
- Adds reboot-interval-in-days flag that are read on every 25m maintain().
 - Changes the reboot algorithm to evenly distribute reboots after 1x-2x reboot intervals after the last reboot. Reboots are now guaranteed to be scheduled before 2x reboot interval.
 - Removes the possibility of overriding durations from environment, which was almost never used.

hakonhall committed:
Allow overriding NodeRepositoryMaintenance durations with flag

freva commented on flags/src/main/java/com/yahoo/vespa/flags/custom/NodeMaintainerDurations.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * {@code NodeRepositoryMaintenance} depends on various durations, e.g. the reboot interval. This class
+ * defines a serialization format for specifying such durations as a JSON object with keys matching
+ * the name (reboot_interval) and with the value being the duration in second as an integer.
```
Let the names be the same as class names? That's what we use to enable/disable maintainers.

hakonhall replied:
The current names are fairly descriptive, while attaching a duration to a Maintainer's subclass name may not be. Also, a Maintainer's interval may be different from the duration controlled here - for instance making smaller adjustments to fail_grace in main would typically not affect the NodeFailer's Maintainer interval.

hakonhall replied:
NA

freva commented on flags/src/main/java/com/yahoo/vespa/flags/custom/NodeMaintainerDurations.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author hakonhall
+ */
+public class NodeMaintainerDurations {
```
Add `hashCode()` and `equals()`

hakonhall replied:
NA

hakonhall committed:
Make flag only for NodeRebooter, and remove fetching from environment

freva reviewed (approved):
LGTM, but I think it would be better to get the flag value in `NodeRebooter`, thereby avoiding having to restart for this to take effect. The interval is in practice overwritten to 25 minutes anyway.

hakonhall committed:
Read reboot-interval-in-days dynamically

But also:

Changes the distribution of the scheduling past 1x reboot interval: hosts will
be scheduled for reboot evenly distributed in the whole 1x-2x range, and is by
this guaranteed to be scheduled at latest at 2x.

The expected time before a reboot was scheduled was 1.33 reboot intervals,
while there was no guarantee of an upper time. The new algorithm has an
expected time before reboot of 1.5 reboot intervals, bound to 2x. The old would
have a higher probability of reboot passing the 1x boundary, while a lower
probability than the new as one nears 2x.

So I think the new algorithm also have the nice property of avoiding thundering
herd, perhaps even more so than the old: For instance when most hosts are
rebooted at the same time in a zone, they would tend to be rescheduled for
reboot closer to each other with the old than with the new.

And, enabling the new algoritm should also not lead to too many hosts suddenly
having to reboot, or at least that's what I hope. I can sanity-check this
before merge - I guess it would be dominated by the number of hosts in
west/east that are beyond 2x.

hakonhall commented:
Done

hakonhall committed:
Do not use discrete probabilities

freva reviewed (approved):
:1st_place_medal:

mpolden commented on node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeRebooter.java:
```diff
@@ -49,13 +52,32 @@ protected void maintain() {
+        // as long as 0 <= overdue <= rebootInterval, with the last maintain() in that interval
+        // naturally scheduling the remaining with probability 1.
+
+        int configServers = 3;
```
Should be determined automatically by using `CuratorDatabaseClient#cluster`.

hakonhall replied:
Fixed in #11313
